sqlitecluster/SQLiteNode.cpp

```diff
@@ -1775,6 +1775,36 @@ void SQLiteNode::_onDisconnect(Peer* peer) {
         _syncPeer = nullptr;
         _changeState(SEARCHING);
     }
+
+    // If we're master, but we've lost quorum, we can't commit anything, nor can worker threads. We need to drop out of
+    // a state that implies we can perform commits, and cancel any outstanding commits.
+    if (_state == MASTERING || _state == STANDINGUP || _state == STANDINGDOWN) {
+        int numFullPeers = 0;
+        int numLoggedInFullPeers = 0;
```

Comment:
I think we already calculate both of these numbers in `bool SQLiteNode::update()` -- it seems like this logic might fit better in these two places:

https://github.com/Expensify/Bedrock/blob/7c95c15b9af0e68a4e9162ad72582f6bcd808dd5/sqlitecluster/SQLiteNode.cpp#L610

https://github.com/Expensify/Bedrock/blob/7c95c15b9af0e68a4e9162ad72582f6bcd808dd5/sqlitecluster/SQLiteNode.cpp#L676

Granted, then it's in two places, but that's where the other state management code is so it's where I would have expected this check to be as well.

Reply:
Yeah, these same numbers are also calculated in `update()`. We can't wait around until `update()` is called to re-calculate them, though, because we need workers to stop committing *immediately*. I thought about refactoring this out into a separate function, but the code didn't seem that much clearer, and it added the possibility of introducing bugs in `update()`, though it would be nice if we didn't repeat ourselves here.

Even with this change, there's still the possible race condition where we commit a transaction after being disconnected from the rest of the cluster. I don't think it's possible to avoid this if we allow ASYNC commits. This change makes the window where that can happen very narrow, however.

Reply:
Regarding "we can't wait around until `update()`", I'm still not sure I agree.  The sync thread loop has very little code between `onPoll()` (which calls `onDisconnect()`) and the next call to `update()` (where the other "change state based on peer connectivity" code exists):

```
        // Process any network traffic that happened.
        syncNode.postPoll(fdm, nextActivity);
        syncNodeQueuedCommands.postPoll(fdm);
        completedCommands.postPoll(fdm);

        // If any of our plugins finished any outstanding HTTPS requests, we'll move those commands back into the
        // regular queue. This code modifies a list while iterating over it.
        auto httpsIt = httpsCommands.begin();
        while (httpsIt != httpsCommands.end()) {
            if (httpsIt->httpsRequest->response) {
                syncNodeQueuedCommands.push(move(*httpsIt));
                httpsIt = httpsCommands.erase(httpsIt);
            } else {
                httpsIt++;
            }
        }

        // Ok, let the sync node to it's updating for as many iterations as it requires. We'll update the replication
        // state when it's finished.
        SQLiteNode::State preUpdateState = syncNode.getState();
        while (syncNode.update()) {}
        SQLiteNode::State nodeState = syncNode.getState();
        replicationState.store(nodeState);
        masterVersion.store(syncNode.getMasterVersion());
```

The only CPU intensive stuff that could add meaningful delay is inside `update()` I think, so it feels a matter of micro/miliseconds between when we detect a failed peer connection and call `onDisconnect()` versus waiting for the next `update()` to find the same thing -- even if under heavy load.

Anyway, I'd still recommend putting it in `update()` just to consolidate that kind of logic in the same place, even despite it extending the potential window by a few ms.  The problem isn't the master continuing on for a few ms longer, but minutes longer, and it feels like it'll work virtually identically in both places so it's better to keep it in `update()`.

That said, If this hasn't convinced you, I don't think it's important enough to debate further.  I'm fine either way, just wanted to clarify this feedback.